drivers/adc/adc_stm32.c

```diff
@@ -610,6 +705,45 @@ static void adc_stm32_teardown_channels(const struct device *dev)
 	adc_stm32_enable(adc);
 }
 
+#ifdef CONFIG_ADC_STM32_DMA
+static void dma_callback(const struct device *dev, void *user_data,
+			 uint32_t channel, int status)
+{
+	/* user_data directly holds the adc device */
+	struct adc_stm32_data *data = user_data;
+	const struct adc_stm32_cfg *config = dev->config;
+	ADC_TypeDef *adc = (ADC_TypeDef *)config->base;
+
+	LOG_DBG("dma callback");
+
+	if (channel == data->dma.channel) {
+		/* Invalidate cache to update buffer.
+		 * Expects buffer address to be 32-byte aligned
+		 * Will invalidate cache at 32-byte granularity
+		 */
+		SCB_InvalidateDCache_by_Addr(data->buffer, data->channel_count * sizeof(int16_t));
+
+		if (LL_ADC_IsActiveFlag_OVR(adc) || (status == 0)) {
+			data->samples_count = data->channel_count;
```

Comment:
```suggestion
			data->samples_count = data->channel_count;
			data->buffer += data->channel_count;
```

I tested this PR on an STM32H745ZI-Q-NUCLEO with extra sampling enabled via `sequence->options->extra_samplings` and noticed that only the first block of samples was written, the rest of the buffer was untouched. Incrementing the buffer pointer here matches what is done in `adc_stm32_isr()` and it made `extra_samplings` write to the whole buffer.

Reply:
Nice catch! Also fixed the unit tests, because it should've failed with this missing.